python/cudf/cudf/core/join/casting_logic.py

```diff
@@ -0,0 +1,282 @@
+                    "unordered categorical variable is ambiguous."
+                )
+        else:
+            if ltype.ordered or rtype.ordered:
+                raise TypeError(
+                    "Inner merge between categoricals with "
+                    "different categories is only valid when "
+                    "neither side is ordered"
+                )
+
+        if not (ltype.ordered or rtype.ordered):
+            # neiter ordered, so categories must be different
+            return _input_to_libcudf_castrules_any(
+                ltype.categories, rtype.categories, how
+            )
+        else:
+            # only one of the operands is ordered
+            raise TypeError(
+                "inner merging on categorical variables when"
+                "only one side is ordered is ambiguous"
+            )
+
+    elif how in {"left", "right"}:
+        if how == "left":
+            major, minor = ltype, rtype
+        elif how == "right":
+            major, minor = rtype, ltype
+
+        # preserve ordering from the major table
+        if minor.ordered and not major.ordered:
+            warnings.warn(
+                f"{how} join does not preserve ordering from"
+                " the minor operand"
+            )
+            if major.categories.equals(minor.categories):
+                return major
+            else:
+                raise TypeError(
+                    f"{how} join when {how} table is unordered"
+                    " and the other is not may only proceed when"
+                    " categories are identical"
+                )
+        elif major.ordered and not minor.ordered:
+            return major
+        elif major.ordered and minor.ordered:
+            # categories must be different
+            raise TypeError(
+                f"{how} join when the {how} categorical"
+                " is unordered, but the other is not, is"
+                " ambiguous"
+            )
+        else:
+            # neither ordered, categories different
+            return major
+
+    elif how == "outer":
+        if (ltype.ordered != rtype.ordered) or (
+            ltype.ordered and rtype.ordered
+        ):
+            # One or both ordered
+            raise TypeError(
+                "Outer join may only proceed when neither"
+                " categorical variable is ordered"
+            )
+        else:
+            # neither ordered, so categories must be different
+            new_cats = cudf.concat(
+                [ltype.categories, rtype.categories]
+            ).unique()
+            return cudf.CategoricalDtype(categories=new_cats, ordered=False)
+
+
+def _input_to_libcudf_castrules_one_cat(lcol, rcol, how):
+    l_is_cat = isinstance(lcol.dtype, CategoricalDtype)
+    r_is_cat = isinstance(rcol.dtype, CategoricalDtype)
+
+    if l_is_cat:
+        return lcol.dtype.categories.dtype
+    elif r_is_cat:
+        return rcol.dtype.categories.dtype
```

Comment:
There's the remote possibility of this function returning `None` in the future. Can we just make this a one-liner inside `_input_to_libcudf_castrules_any_cat` below?

```
return lcol.dtype.categories.dtype if l_is_cat else r_is_cat
```

Reply:
Done